extensions/azurecore/src/account-provider/azureAccountProvider2.ts

```diff
@@ -0,0 +1,443 @@
+		}
+	}
+
+	private addServerPaths(
+		pathMappings: Map<string, (req: http.IncomingMessage, res: http.ServerResponse, reqUrl: url.UrlWithParsedQuery) => void>,
+		nonce: string,
+		authUrl: string) {
+
+		const initialSignIn = ((req: http.IncomingMessage, res: http.ServerResponse, reqUrl: url.UrlWithParsedQuery) => {
+			const receivedNonce = (reqUrl.query.nonce as string || '').replace(/ /g, '+');
+			if (receivedNonce !== nonce) {
+				// TODO handle broken nonce situation
+				return;
+			}
+			res.writeHead(302, { Location: authUrl });
+			res.end();
+		});
+
+		const callback = ((req: http.IncomingMessage, res: http.ServerResponse, reqUrl: url.UrlWithParsedQuery) => {
+			const state = reqUrl.query.state as string ?? '';
+			const code = reqUrl.query.code as string ?? '';
+
+			const stateSplit = state.split(',');
+			if (stateSplit.length !== 2) {
+				// TODO handle broken state.
+				return;
+			}
+
+			if (stateSplit[1] !== nonce) {
+				// TODO handle broken nonce situation
+				return;
+			}
+
+			res.writeHead(200, 'success, you may close this page');
+			res.end();
+
+			this.handleAuthentication(code).catch(console.error);
+		});
+
+		pathMappings.set('/signin', initialSignIn);
+		pathMappings.set('/callback', callback);
+	}
+	private async makeWebRequest(accessToken: TokenResponse, uri: string): Promise<any> {
+		const params = {
+			headers: {
+				'Content-Type': 'application/json',
+				'Authorization': `Bearer ${accessToken.accessToken}`
+			},
+			json: true
+		};
+
+		return new Promise((resolve, reject) => {
+			request.get(uri, params, (error: any, response: request.Response, body: any) => {
+				const err = error ?? body.error;
+				if (err) {
+					return reject(err);
+				}
+				return resolve(body.value);
+			});
+		});
+	}
+
+	private async getTenants(userId: string, homeTenant: string): Promise<Tenant[]> {
+		const armToken = await this.getToken(userId, AzureAccountProvider.AadCommonTenant, this.metadata.settings.armResource.id);
+		const tenantUri = url.resolve(this.metadata.settings.armResource.endpoint, 'tenants?api-version=2015-01-01');
+		const armWebResponse: any[] = await this.makeWebRequest(armToken, tenantUri);
+
+		const promises = armWebResponse.map(async (value: { tenantId: string }) => {
+			const graphToken = await this.getToken(userId, value.tenantId, this.metadata.settings.graphResource.id);
+			let tenantDetailsUri = url.resolve(this.metadata.settings.graphResource.endpoint, value.tenantId + '/');
+			tenantDetailsUri = url.resolve(tenantDetailsUri, 'tenantDetails?api-version=2013-04-05');
+			const tenantDetails: any[] = await this.makeWebRequest(graphToken, tenantDetailsUri);
+
+			return {
+				id: value.tenantId,
+				userId: userId,
+				displayName: tenantDetailsUri.length && tenantDetails[0].displayName
+					? tenantDetails[0].displayName
+					: localize('azureWorkAccountDisplayName', "Work or school account")
+			} as Tenant;
+		});
+
+		const tenants = await Promise.all(promises);
```

Comment:
If an error occurs in one of the promises we won't execute the below code - is that intentional?

Reply:
Yeah, if an error occurs there we should stop what we're doing.